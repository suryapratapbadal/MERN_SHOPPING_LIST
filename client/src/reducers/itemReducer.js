
const initialState = {
    items: [],
    recipes: [],
    loading: false,
    user: false,
};

const itemReducer = (state = initialState, action) => {
    const { type, payload } = action;
    switch (type) {
        case 'GET_ITEMS':
            return {
                ...state,
                items: payload,
            };

        case 'ADD_ITEM':
            return {
                ...state,
                items: [...state.items, payload],
                loading: false
            };

        case 'DELETE_ITEM':
            return {
                ...state,
                items: state.items.filter(item => item._id !== payload),
                loading: false
            };

        case 'UPDATE_ITEM':
            return {
                ...state,
                items:[...state.items.filter(item => item._id !== payload._id), payload],
                loading: false
            };

        case 'VERIFY_USER':
            return {
                ...state,
                user: true
            };

        case 'LOG_OUT_USER':
            return {
                ...state,
                user: false
            };

        case 'GET_RECIPES':
            return {
                ...state,
                recipes: payload,
            };

        case 'ADD_RECIPE':
            return {
                ...state,
                recipes: state.recipes ? [...state.recipes, payload] : [],
                loading: false
            };

        case 'DELETE_RECIPE':
            return {
                ...state,
                recipes: state.recipes.filter(recipes => recipes._id !== payload),
                loading: false
            };

        case 'LOADING':
            return {
                ...state,
                loading: true,
            };

        default:
            return { ...state };
    }
};

export default itemReducer;