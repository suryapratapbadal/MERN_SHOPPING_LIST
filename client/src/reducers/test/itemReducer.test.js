import itemReducer from '../itemReducer';


describe('select_reducer', () => {
  describe('INITIAL_STATE', () => {
    test('is correct', () => {
      const action = { type: 'dummy_action' };
      const initialState = { items: [], loading: false, recipes: [], user: true };

      expect(itemReducer(undefined, action)).toEqual(initialState);
      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('GET_ITEMS', () => {
    test('returns the correct state', () => {
      const action = { type: 'GET_ITEMS', payload: { items: [{ id: 1, name: 'Milk' }] } };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('ADD_ITEM', () => {
    test('returns the correct state', () => {
      const action = { type: 'ADD_ITEM', payload: { items: [{ id: 1, name: 'Milk' }] } };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('DELETE_ITEM', () => {
    test('returns the correct state', () => {
      const action = { type: 'DELETE_ITEM', payload: 1 };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('UPDATE_ITEM', () => {
    test('returns the correct state', () => {
      const action = { type: 'UPDATE_ITEM', payload: { items: [{ id: 1, name: 'Milk' }] }  };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('VERIFY_USER', () => {
    test('returns the correct state', () => {
      const action = { type: 'VERIFY_USER', payload: '' };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('LOG_OUT_USER', () => {
    test('returns the correct state', () => {
      const action = { type: 'LOG_OUT_USER', payload: '' };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('GET_RECIPES', () => {
    test('returns the correct state', () => {
      const action = { type: 'GET_RECIPES', payload: { recipes: [{ id: 1, name: 'tea' }] } };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('ADD_RECIPE', () => {
    test('returns the correct state', () => {
      const action = { type: 'ADD_RECIPE', payload: { recipes: [{ id: 1, name: 'tea' }] } };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('DELETE_RECIPE', () => {
    test('returns the correct state', () => {
      const action = { type: 'DELETE_RECIPE', payload: 1 };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('LOADING', () => {
    test('returns the correct state', () => {
      const action = { type: 'LOADING', payload: '' };

      expect(itemReducer(undefined, action)).toMatchSnapshot();
    });
  });
});

