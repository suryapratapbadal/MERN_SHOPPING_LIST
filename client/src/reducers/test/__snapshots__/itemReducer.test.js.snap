// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select_reducer ADD_ITEM returns the correct state 1`] = `
Object {
  "items": Array [
    Object {
      "items": Array [
        Object {
          "id": 1,
          "name": "Milk",
        },
      ],
    },
  ],
  "loading": false,
  "recipes": Array [],
  "user": true,
}
`;

exports[`select_reducer ADD_RECIPE returns the correct state 1`] = `
Object {
  "items": Array [],
  "loading": false,
  "recipes": Array [
    Object {
      "recipes": Array [
        Object {
          "id": 1,
          "name": "tea",
        },
      ],
    },
  ],
  "user": true,
}
`;

exports[`select_reducer DELETE_ITEM returns the correct state 1`] = `
Object {
  "items": Array [],
  "loading": false,
  "recipes": Array [],
  "user": true,
}
`;

exports[`select_reducer DELETE_RECIPE returns the correct state 1`] = `
Object {
  "items": Array [],
  "loading": false,
  "recipes": Array [],
  "user": true,
}
`;

exports[`select_reducer GET_ITEMS returns the correct state 1`] = `
Object {
  "items": Object {
    "items": Array [
      Object {
        "id": 1,
        "name": "Milk",
      },
    ],
  },
  "loading": false,
  "recipes": Array [],
  "user": true,
}
`;

exports[`select_reducer GET_RECIPES returns the correct state 1`] = `
Object {
  "items": Array [],
  "loading": false,
  "recipes": Object {
    "recipes": Array [
      Object {
        "id": 1,
        "name": "tea",
      },
    ],
  },
  "user": true,
}
`;

exports[`select_reducer INITIAL_STATE is correct 1`] = `
Object {
  "items": Array [],
  "loading": false,
  "recipes": Array [],
  "user": true,
}
`;

exports[`select_reducer LOADING returns the correct state 1`] = `
Object {
  "items": Array [],
  "loading": true,
  "recipes": Array [],
  "user": true,
}
`;

exports[`select_reducer LOG_OUT_USER returns the correct state 1`] = `
Object {
  "items": Array [],
  "loading": false,
  "recipes": Array [],
  "user": false,
}
`;

exports[`select_reducer UPDATE_ITEM returns the correct state 1`] = `
Object {
  "items": Array [
    Object {
      "items": Array [
        Object {
          "id": 1,
          "name": "Milk",
        },
      ],
    },
  ],
  "loading": false,
  "recipes": Array [],
  "user": true,
}
`;

exports[`select_reducer VERIFY_USER returns the correct state 1`] = `
Object {
  "items": Array [],
  "loading": false,
  "recipes": Array [],
  "user": true,
}
`;
